test/test_type_promotion.py

```diff
@@ -430,6 +431,97 @@ def test_transpose(self, device):
         a = torch.tensor([[True, True], [False, True]], device=device)
         self.assertEqual(a.t() == 0, a.t() == False)  # noqa: E712
 
+    def get_sparse_tensors(self, device, dtype, coalesced, zeros=True):
+        t = self._get_test_tensor(device, dtype, not zeros)
+        if coalesced:
+            s = t.to_sparse()
+        else:
+            s = t.to_sparse()
+            indices = torch.cat((s.indices(), s.indices()), 1)
+            values = torch.cat((s.values(), s.values()), 0)
+            s = torch.sparse_coo_tensor(indices=indices, values=values, size=s.size(), dtype=dtype, device=device)
+            t = s.to_dense()
+        return (t, s)
+
+    def _get_precision(self, dtype, coalesced):
+        if dtype == torch.half and not coalesced:
+            # very low precision for uncoalesced float16 sparse tensors since
```

Comment:
This seems bad

Reply:
yeah that's just a fact of low precision though:

```
In [20]: values=[10.98764] * 4
In [21]: values
Out[21]: [10.98764, 10.98764, 10.98764, 10.98764]
In [22]: torch.tensor(values, dtype=torch.half, device='cuda').sum() - torch.tensor(values, dtype=torch.float).sum()
Out[22]: tensor(-0.0131, device='cuda:0')
```